blockstore/splitstore/splitstore_prune.go

```diff
@@ -0,0 +1,592 @@
+	s.compactType = cold
+	go func() {
+		defer atomic.StoreInt32(&s.compacting, 0)
+		defer s.endTxnProtect()
+
+		log.Info("pruning splitstore")
+		start := time.Now()
+
+		s.prune(curTs, retainStateP, doGC)
+
+		log.Infow("prune done", "took", time.Since(start))
+	}()
+
+	return nil
+}
+
+func (s *SplitStore) prune(curTs *types.TipSet, retainStateP func(int64) bool, doGC func() error) {
+	log.Info("waiting for active views to complete")
+	start := time.Now()
+	s.viewWait()
+	log.Infow("waiting for active views done", "took", time.Since(start))
+
+	err := s.doPrune(curTs, retainStateP, doGC)
+	if err != nil {
+		log.Errorf("PRUNE ERROR: %s", err)
+	}
+}
+
+func (s *SplitStore) doPrune(curTs *types.TipSet, retainStateP func(int64) bool, doGC func() error) error {
+	currentEpoch := curTs.Height()
+	log.Infow("running prune", "currentEpoch", currentEpoch, "baseEpoch", s.baseEpoch)
+
+	markSet, err := s.markSetEnv.New("live", s.markSetSize)
+	if err != nil {
+		return xerrors.Errorf("error creating mark set: %w", err)
+	}
+	defer markSet.Close() //nolint:errcheck
+	defer s.debug.Flush()
+
+	if err := s.checkClosing(); err != nil {
+		return err
+	}
+
+	// 0. track all protected references at beginning of compaction; anything added later should
+	//    be transactionally protected by the write
+	log.Info("protecting references with registered protectors")
+	err = s.applyProtectors()
+	if err != nil {
+		return err
+	}
+
+	// 1. mark reachable objects by walking the chain from the current epoch; we keep all messages
+	//    and chain headers; state and reciepts are retained only if it is within retention policy scope
+	log.Info("marking reachable objects")
+	startMark := time.Now()
+
+	count := new(int64)
+	err = s.walkChainDeep(curTs, retainStateP,
+		func(c cid.Cid) error {
+			if isUnitaryObject(c) {
+				return errStopWalk
+			}
+
+			mark, err := markSet.Has(c)
+			if err != nil {
+				return xerrors.Errorf("error checking markset: %w", err)
+			}
+
+			if mark {
+				return errStopWalk
+			}
+
+			atomic.AddInt64(count, 1)
+			return markSet.Mark(c)
+		})
+
+	if err != nil {
+		return xerrors.Errorf("error marking: %w", err)
+	}
+
+	log.Infow("marking done", "took", time.Since(startMark), "marked", count)
+
+	if err := s.checkClosing(); err != nil {
+		return err
+	}
+
+	// 1.1 protect transactional refs
+	err = s.protectTxnRefs(markSet)
+	if err != nil {
+		return xerrors.Errorf("error protecting transactional refs: %w", err)
+	}
+
+	if err := s.checkClosing(); err != nil {
+		return err
+	}
+
+	// 2. iterate through the coldstore to collect dead objects
+	log.Info("collecting dead objects")
+	startCollect := time.Now()
+
+	deadw, err := NewColdSetWriter(s.deadSetPath())
+	if err != nil {
+		return xerrors.Errorf("error creating coldset: %w", err)
+	}
+	defer deadw.Close() //nolint:errcheck
+
+	// some stats for logging
+	var liveCnt, deadCnt int
+
+	err = s.cold.(bstore.BlockstoreIterator).ForEachKey(func(c cid.Cid) error {
```

Comment:
This feels like it will be very very slow on larger nodes (which is maybe fine, just don't do coldstore gc on archival nodes)

Reply:
Do you think this applies to nodes running with historical message data?  I've been hoping to make the lotus default do cold store pruning indefinitely and keep growing message datastore size.  Maybe I need to rethink this?

Reply:
Yeah, we probably want some metrics around this.

Also doing marking/gc in batches instead of having to mark all dead objects first could help a lot - now you can at least restart your node wihdout losing that much progress, and things just get lighter at their own pace.

Not sure if we want this in this PR, possibly just saying that splitstore may not work that well with really big coldstores is good enough.

Reply:
I kicked off some brainstorming here: https://github.com/filecoin-project/lotus/discussions/9128. We have a few good options if this starts to look like a problem experimentally.